File: Java/84. LargestRectangleInHistogram.java
// Explaination - for any index i we will take help of array nser and array nsel to find the next smaller element to the right and next smaller element to the left
// respectively. One we get those tho value for a particular index i, we will find the width as nser[i]-nsel[i]-1; Since we have the height already as heights[i]
// widht, we can find the area;
// Prerequisite - Next greater/smaller element on the left/right;
class Solution {
    public int largestRectangleArea(int[] heights) {
        if(heights.length == 0){
            return 0;
        }
        int[] nser = new int[heights.length]; // array contains next smaller element to the right
        Stack<Integer> st = new Stack<>();
        
        st.push(heights.length-1);
        nser[heights.length-1] = heights.length;
        
        for(int i=heights.length-2; i>=0; i--){
            while(st.size()>0 && heights[i]<=heights[st.peek()]){
                st.pop();
            }
            
            if(st.size() == 0){
                nser[i] = heights.length;
            }else{
                nser[i] = st.peek();
            }
            st.push(i);
        }
        
        int[] nsel = new int[heights.length]; // array contains next smaller element to the left
        st = new Stack<>();
        
        st.push(0);
        nsel[0] = -1;
        
        for(int i=1; i<heights.length; i++){
            while(st.size()>0 && heights[i]<=heights[st.peek()]){
                st.pop();
            }
            
            if(st.size() == 0){
                nsel[i] = -1;
            }else{
                nsel[i] = st.peek();
            }
            st.push(i);
        }
        
        int maxArea = 0;
        for(int i=0; i<heights.length; i++){
            int width = nser[i]-nsel[i]-1;
            int area = width*heights[i];
            if(area>maxArea){
                maxArea = area;
            }
        }
        
        return maxArea;
    }
}

File: C++/84.LargestRectangleInHistogram.cpp
//Problem 84.Largest Rectangle in Histogram

//CODE
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
       int n = heights.size();
        int max_area = 0;
		//We maintain a stack which stores the index of the shortest rectangle encountered so far and those of consequent
		//rectangles taller than it. As we encounter shorter rectangles than the tallest rectangle in the stack,  we pop elements
		//to maintain a  stack with indices of rectangles with non-decreasing heights. The key idea is that a rectangle with 
		//height equal to that of the 'minimum encountered so far' can extend into subsequent rectangles. As we pop an 
		//indices from the stack, we need to record the maximum area that a rectangle with the poped rectangle's height could
		//occupy which is simply the index until (not including) i in the for loop till (not including) the previously unpoped 
		//element present in the stack.
        stack<int> s;
        for (int i=0; i<heights.size(); ++i) {
            while (!s.empty() && heights[s.top()] > heights[i]) {
                int cur_max = s.top();
                s.pop();
                //Add upto previous index -> (i-1) - (the index of the last value we didn't kick out + 1) + 1.
				//If s is empty we are simply considering a recrtangle starting from i=0.
                int area = heights[cur_max] * ((i-1) - ((s.empty()?-1:s.top())+1) + 1);
                max_area = max(area,max_area);
            }
            s.push(i);
        }
        //If the last few values are increasing, we need to remove these values.
        while (!s.empty()) {
            int cur_max = s.top();
            s.pop();
            //Add upto previous index -> (n-1) - (the index of the value we kicked out) + 1.
            int area = heights[cur_max] * ((n-1) - ((s.empty()?-1:s.top())+1) + 1);
            max_area = max(area,max_area);
        }
        
        return max_area;
        
    }
};
